### DISCUSSION TITLE
Mockup SLURM REST API

### DISCUSSION THREAD

**Discussion by JohnBioinf** [mockoon/mockoon#1228] [Help]
Hi everybody

I currently try to setup mockoon for development and testing. The service should mock up a slurm cluster to which my service submits jobs.
The specification for the slurm api can be found [here](https://github.com/SchedMD/slurm/blob/master/src/slurmrestd/plugins/openapi/v0.0.39/openapi.json).

As an end result I would like to start mockoon in a docker container. Currently I try to work things out directly with the mockoon-cli locally.

What I did so far:
```bash
$ mockoon-cli start --data https://github.com/SchedMD/slurm/blob/master/src/slurmrestd/plugins/openapi/v0.0.39/openapi.json
This environment does not seem to be a valid Mockoon environment or is too old. Let Mockoon attempt to repair it? (y/n): y
{"app":"mockoon-server","environmentName":"New environment","environmentUUID":"e357ba4e-7fe2-413f-bcf0-f1ae27e18c5d","level":"info","message":"Server started on port 3000","timestamp":"2023-12-12T10:31:56.130Z"}
{"app":"mockoon-server","environmentName":"New environment","environmentUUID":"e357ba4e-7fe2-413f-bcf0-f1ae27e18c5d","level":"info","message":"Transaction recorded","requestMethod":"GET","requestPath":"/slurm/v0.0.39/ping","requestProxied":false,"responseStatus":404,"timestamp":"2023-12-12T10:33:55.893Z"}

$ curl 'http://localhost:3000/slurm/v0.0.39/ping/'
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /slurm/v0.0.39/ping/</pre>
</body>
</html>
```

Why does the service not find the ping path?

Snippet from json:
```yml
paths:
  /ping/:
    get:
      tags:
        - slurm
      operationId: ping
      summary: ping test
      responses:
        '200':
          description: results of ping test
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/v0.0.39_pings'
            application/x-yaml:
              schema:
                $ref: '#/components/schemas/v0.0.39_pings'
        default:
          description: unable to request ping test
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/status'
            application/x-yaml:
              schema:
                $ref: '#/components/schemas/status'
```

**Comment by 255kb** [comment]
Hi, 

First of all, I don't think you can directly load this URL, as it is the URL from GitHub's web page. You should use `https://raw.githubusercontent.com/SchedMD/slurm/master/src/slurmrestd/plugins/openapi/v0.0.39/openapi.json` instead. 
Second, we are using the `derefence` method of the https://apitools.dev/swagger-parser/ library. It seems that there is a `$ref` containing constants like `DATA_PARSER_JOB_DESC_MSG` or `DATA_PARSER_LICENSES` which are neither internal references nor files available on their repository. That's why the derefence is failing:
![image](https://github.com/mockoon/mockoon/assets/7489814/0ffa2dd1-1db6-46c4-9d98-2b4c1b7aa54c)

The message "This environment does not seem to be a valid Mockoon environment..." gave a clue as it is the last one you get after everything else failed, including OpenAPI import. But the CLI should have reported better the swagger-parser error. I opened [an issue](https://github.com/mockoon/mockoon/issues/1229) to improve this for the next release. 

I don't think I have a solution for you as this `$ref` seem to be missing, or refering to some internal data. The file also fails [Swagger online validator](https://apitools.dev/swagger-parser/online/). You can try to remove them or replace with simple string.

I did it in the file below (though I'm not an OpenAPI expert, and just replaced everything with a string schema): 
[openapi.json](https://github.com/mockoon/mockoon/files/13648636/openapi.json)

**Reply by JohnBioinf** [reply]
Thanks for the answer. I have had a look at the source code of the slurm rest api. I know little about C and even less about this project but I would guess that this is dynamically compiled depending on the system.
So the best way forward would be to fiugure how they are set. I guess.
But the problem lies in the incomplete openapi file, not in mockoon.

**Reply by 255kb** [reply]
Yes, aside from $refs OpenAPI spec files shouldn't rely on external compilation. Or it should be documented.
